xarray/core/coordinates.py

```diff
@@ -796,6 +797,121 @@ def _ipython_key_completions_(self):
+        return set(self._data._coord_variables)
+
+    @property
+    def dims(self) -> Frozen[Hashable, int]:
+        # TODO is there a potential bug here? What happens if a dim is only present on data variables?
+        return self._data.dims
+
+    @property
+    def dtypes(self) -> Frozen[Hashable, np.dtype]:
+        """Mapping from coordinate names to dtypes.
+
+        Cannot be modified directly, but is updated when adding new variables.
+
+        See Also
+        --------
+        Dataset.dtypes
+        """
+        return Frozen({n: v.dtype for n, v in self._data._coord_variables.items()})
+
+    @property
+    def variables(self) -> Mapping[Hashable, Variable]:
+        return Frozen(self._data._coord_variables)
+
+    def __getitem__(self, key: Hashable) -> DataArray:
+        if key not in self._data._coord_variables:
+            raise KeyError(key)
+        return self._data[key]
+
+    def to_dataset(self) -> Dataset:
+        """Convert these coordinates into a new Dataset"""
+        return self._data.to_dataset()._copy_listed(self._names)
+
+    def _update_coords(
+        self, coords: dict[Hashable, Variable], indexes: Mapping[Any, Index]
+    ) -> None:
+
+        # TODO I don't know how to update coordinates that live in parent nodes
+        # TODO We would have to find the correct node and update `._node_coord_variables`
+
+        # create updated node (`.to_dataset` makes a copy so this doesn't modify in-place)
+        node_ds = self._data.to_dataset(inherited=False)
+        node_ds.coords._update_coords(coords, indexes)
+
+        from xarray.core.datatree import _check_alignment
+
+        # check consistency *before* modifying anything in-place
+        # TODO can we clean up the signature of _check_alignment to make this less awkward?
+        if self._data.parent is not None:
+            parent_ds = self._data.parent._to_dataset_view(rebuild_dims=False)
+        else:
+            parent_ds = None
+        _check_alignment(self._data.path, node_ds, parent_ds, self._data.children)
+
+        # assign updated attributes
+        coord_variables = {
+            k: v for k, v in node_ds.variables.items() if k in node_ds._coord_names
+        }
+        self._data._node_coord_variables = coord_variables
+        self._data._node_dims = node_ds._dims
+        self._data._node_indexes = node_ds._indexes
+
+    def _drop_coords(self, coord_names):
+        # should drop indexed coordinates only
+        for name in coord_names:
+            del self._data._node_coord_variables[name]
+            del self._data._node_indexes[name]
+
+    def _drop_indexed_coords(self, coords_to_drop: set[Hashable]) -> None:
+        assert self._data.xindexes is not None
+        new_coords = drop_indexed_coords(coords_to_drop, self)
+        for name in self._data._node_coord_variables - new_coords._names:
+            del self._data._node_coord_variables[name]
+        self._data._node_indexes = dict(new_coords.xindexes)
```

Comment:
I'm pretty sure that this method, and its equivalent in `DatasetCoordinates`, is never being called... I'm about to insert `raise` in both methods and pass the whole test suite.

@dcherian @benbovy you both touched this method last, do you know why it still exists?

Reply:
Removed in https://github.com/pydata/xarray/pull/9451/commits/9dc845a674bf1090b76589dd69a15ec42869ccc1